cabal-install/Distribution/Solver/Modular/Explore.hs

```diff
@@ -40,57 +45,101 @@ import qualified Distribution.Solver.Types.Progress as P
 -- with the (virtual) option not to choose anything for the current
 -- variable. See also the comments for 'avoidSet'.
 --
-backjump :: F.Foldable t => EnableBackjumping -> Var QPN
-         -> ConflictSet QPN -> t (ConflictSetLog a) -> ConflictSetLog a
+backjump :: EnableBackjumping -> Var QPN
+         -> ConflictSet QPN -> P.PSQ k (Explore (ConflictSetLog a))
+         -> Explore (ConflictSetLog a)
 backjump (EnableBackjumping enableBj) var initial xs =
     F.foldr combine logBackjump xs initial
   where
-    combine :: ConflictSetLog a
-            -> (ConflictSet QPN -> ConflictSetLog a)
-            ->  ConflictSet QPN -> ConflictSetLog a
-    combine (P.Done x)    _ _               = P.Done x
-    combine (P.Fail cs)   f csAcc
-      | enableBj && not (var `CS.member` cs) = logBackjump cs
-      | otherwise                = f (csAcc `CS.union` cs)
-    combine (P.Step m ms) f cs   = P.Step m (combine ms f cs)
+    combine ::  Explore (ConflictSetLog a)
+            -> (ConflictSet QPN -> Explore (ConflictSetLog a))
+            ->  ConflictSet QPN -> Explore (ConflictSetLog a)
+    combine x f csAcc = do
+      l <- x
+      case l of
+        P.Done d  -> return (P.Done d)
+        P.Fail cs
+          | enableBj && not (var `CS.member` cs) -> logBackjump cs
+          | otherwise                            -> f (csAcc `CS.union` cs)
+        P.Step m ms -> do
+          l' <- combine (return ms) f csAcc
+          return (P.Step m l')
 
-    logBackjump :: ConflictSet QPN -> ConflictSetLog a
-    logBackjump cs = failWith (Failure cs Backjump) cs
+    logBackjump :: ConflictSet QPN -> Explore (ConflictSetLog a)
+    logBackjump cs = return (failWith (Failure cs Backjump) cs)
 
 type ConflictSetLog = P.Progress Message (ConflictSet QPN)
 
+type ConflictMap = Map (Var QPN) Int
+
+getBestGoal :: ConflictMap -> P.PSQ (Goal QPN) a -> (Goal QPN, a)
+getBestGoal cm =
+  P.maximumBy
+    ( flip (M.findWithDefault 0) cm
+    . (\ (Goal v _) -> v)
+    )
+
+getFirstGoal :: P.PSQ (Goal QPN) a -> (Goal QPN, a)
+getFirstGoal ts =
+  P.casePSQ ts
+    (error "getFirstGoal: empty goal choice") -- empty goal choice is an internal error
+    (\ k v _xs -> (k, v))  -- commit to the first goal choice
+
+updateCM :: ConflictSet QPN -> ConflictMap -> ConflictMap
+updateCM cs cm =
+  L.foldl' (\ cmc k -> M.alter inc k cmc) cm (CS.toList cs)
```

Comment:
Yes. But `insertWith'` is deprecated in containers-0.5. We could just use a strict finite map, but `cabal-install` currently only specified `containers >= 0.4`. On the other hand, there seem to be other modules in `cabal-install` that make use of `Data.Map.Strict`, but with CPP directives to be backwards-compatible. Taking this into account, I am not sure that the change is worth it, because the use of `alter` seems to be neither deprecated nor require adaptation between different versions of the `containers` package. @23Skidoo, do you have any opinion on this?